File: CosmeticaShop.Web/Areas/Admin/assets/js/Form/Category/CategoryModel.js
var Category = Category || {};

(function () {

    if (Category.CategoryModel) {
        console.error("Category.CategoryModel уже был создан");
        return;
    }

    /**
     * Класс для описания модели категории
     * @param {     
     * } theParams 
     * @returns {} 
     */
    Category.CategoryModel = function (theParams) {
        theParams = theParams || {};

        this.Id = ko.observable(theParams.Id || "");
        this.Name = ko.observable(theParams.Name || "");
        this.ParentId = ko.observable(theParams.ParentId || "");
        this.KeyUrl = ko.observable(theParams.KeyUrl || "");

        this.IsActive = ko.observable(theParams.IsActive || false);

        this.IsClosed = ko.observable(false);

        this.ChildCategories = ko.observableArray((theParams.ChildCategories || [])
            .map(function (item) { return new Category.CategoryModel(item) }));
        return this;
    };

    /**
     * Определяем конструктор
     */
    Category.CategoryModel.prototype.constructor = Category.CategoryModel;

    Category.CategoryModel.prototype.log = function (text) {
        console.log("Ошибка в классе Category.CategoryModel: " + text);
    };

    Category.CategoryModel.prototype.GetData = function () {
        return {
            Id: this.Id(),
            Name: this.Name(),
            ParentId: this.ParentId()===0?"":this.ParentId(),
            KeyUrl: this.KeyUrl(),
            IsActive: this.IsActive()
        }
    }
})();

File: CosmeticaShop.IServices/Models/Category/CategoryModel.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Web;
using Newtonsoft.Json;

namespace CosmeticaShop.IServices.Models.Category
{
    public class CategoryModel
    {
        #region [ Свойства ]

        /// <summary>
        /// Ид
        /// </summary>
        public int Id { get; set; }

        /// <summary>
        /// Ид родителя
        /// </summary>
        public int? ParentId { get; set; }

        /// <summary>
        /// Наименование
        /// </summary>
        public string Name { get; set; }

        /// <summary>
        /// Url категории
        /// </summary>
        public string KeyUrl { get; set; }

        /// <summary>
        /// Приоритет
        /// </summary>
        public int? Priority { get; set; }

        /// <summary>
        /// Видимость
        /// </summary>
        public bool IsActive { get; set; }
        /// <summary>
        /// Родительская категория
        /// </summary>
        [JsonIgnore]
        public CategoryModel Parent { get; set; }
        /// <summary>
        /// Список дочерних категорий
        /// </summary>
        public List<CategoryModel> ChildCategories { get; set; }

        #endregion
    }
}
